framework/src/hdgkernels/HDGKernel.C

```diff
@@ -0,0 +1,176 @@
+      if (_hibc_warehouse.hasActiveBoundaryObjects(bnd_id, _tid))
+      {
+        _fe_problem.setCurrentBoundaryID(bnd_id, _tid);
+
+        if (!fe_face_reinitd)
+        {
+          fe_face_reinitd = true;
+          // Do the whole shebang:
+          // - reinit FE objects on the face
+          // - compute variable values on the face
+          // - compute material values, including values from face and boundary materials
+          NonlinearThread::prepareFace(_fe_problem, _tid, _current_elem, side, _current_bnd_id);
+        }
+        else
+          // Only need to compute boundary materials; face materials have already been computed on
+          // the current face
+          _fe_problem.reinitMaterialsBoundary(_current_bnd_id, _tid);
+
+        libmesh_assert(_current_side == side);
+        const auto & bcs = _hibc_warehouse.getActiveBoundaryObjects(bnd_id, _tid);
+        for (const auto & bc : bcs)
+        {
+          mooseAssert(bc->shouldApply(),
+                      "I don't think anyone uses the shouldApply feature for integrated boundary "
+                      "conditions");
+          bc->onBoundary();
+          addBCData(*bc);
+        }
+      }
+
+    if (_neigh = _current_elem->neighbor_ptr(side);
+        _neigh && this->hasBlocks(_neigh->subdomain_id()))
+    {
+      if (!fe_face_reinitd)
+      {
+        fe_face_reinitd = true;
+        NonlinearThread::prepareFace(
+            _fe_problem, _tid, _current_elem, side, Moose::INVALID_BOUNDARY_ID);
+      }
+
+      libmesh_assert(_current_side == side);
+      onInternalSide();
+    }
+  }
+
+  _PrimalMatInv = _PrimalMat.inverse();
+  const auto lm_size = _lm_dof_indices.size();
+  if (_computing_global_data)
+  {
+    _K_libmesh.resize(lm_size, lm_size);
+    _F_libmesh.resize(lm_size);
+    const auto LMProductMat = -_LMPrimal * _PrimalMatInv;
+    _LMMat += LMProductMat * _PrimalLM;
+    _LMVec += LMProductMat * _PrimalVec;
+    libmesh_assert(cast_int<std::size_t>(_LMMat.rows()) == lm_size);
+    libmesh_assert(cast_int<std::size_t>(_LMMat.cols()) == lm_size);
+    libmesh_assert(cast_int<std::size_t>(_LMVec.size()) == lm_size);
+
+    for (const auto i : make_range(lm_size))
+    {
+      for (const auto j : make_range(lm_size))
+        _K_libmesh(i, j) = _LMMat(i, j);
+      _F_libmesh(i) = _LMVec(i);
+    }
+
+    // We were performing our finite element assembly for the implicit solve step of our
+    // example. Add our local element vectors/matrices into the global system
+    addResiduals(_assembly, _F_libmesh, _lm_dof_indices, /*scaling_factor=*/1);
+    addJacobian(_assembly, _K_libmesh, _lm_dof_indices, _lm_dof_indices, /*scaling_factor=*/1);
+  }
+  else
+  {
+    //
+    // We are doing our finite element assembly for the second time. We now know the Lagrange
+    // multiplier solution. With that and the local element matrices and vectors we can compute
+    // the vector and scalar solutions
+    //
```

Comment:
so which loop are in now? ComputeElemAux ?
and which loop are we in the first statement of this if condition ? the PreCheck ?

Reply:
I will add a comment. It's the `ComputeResidualAndJacobianThread` for the former and the now `HDGPrimalSolutionUpdateThread` for the latter